### PR TITLE
2124 rewrite bookmarks activity

### LINKED ISSUE #2124: Rewrite bookmarks activity to adhere to unidirectional data flow
**Is your feature request related to a problem? Please describe.**
There are a "host of data flow issues because of their outdated architecture" (HistoryActivity & BookmarksActivity). #1990 <- also contains an example of an issue caused by poor data flow in the HistoryActivity.

**Describe the solution you'd like**
Bookmarks activity should be rewritten to adhere to unidirectional data flow to solve these issues.

**Additional context**
This is a child issue to #1990 .

### PR DESCRIPTION
**Pull request by Frans-Lukas**
Fixes #2124 

This PR rewrites the bookmarks activity to use better design patterns. This PR also contains the HistoryActivity rewrite as I used it for reference, this is only until I have finished the issue or until the HistoryActivity PR has been merged with the Dev branch. 

I haven't quite understood the repository compared to just using e.x. the HistoryDao/BookmarksDao. 
To delete bookmarks I tried:
```
repository.deleteBookmarks(bookmarkItems)
```
But that does not delete the items? Any advice/help is appreciated.

### CODE REVIEW

File: core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/adapter/BookmarkViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import org.kiwix.kiwixmobile.core.extensions.setBitmap
+import org.kiwix.kiwixmobile.core.extensions.setImageDrawableCompat
+
+class BookmarkViewHolder(
+  override val containerView: View,
+  private val itemClickListener: OnItemClickListener
+) : BaseViewHolder<BookmarkItem>(containerView) {
+  override fun bind(item: BookmarkItem) {
+    title.text = item.bookmarkTitle
+    Log.d("BookmarkItemViewHolder", "isSelected=${item.isSelected}")
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/adapter/BookmarkViewHolder.kt:20]: delete this log

File: core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDao.kt [hunk 2]
```diff
@@ -40,6 +40,8 @@ class NewBookmarksDao @Inject constructor(val box: Box<BookmarkEntity>) {
   }.find()
     .map(::BookmarkItem)
 
+  fun bookmarks() = box.asFlowable().map { it.map(::BookmarkItem) }
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDao.kt:43]: I imagine some corresponding thing can be deleted with the addition of this method?

File: core/src/main/res/values-kab/strings.xml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<resources xmlns:tools="http://schemas.android.com/tools" tools:ignore="DuplicateStrings">
```
Comment (macgills) [inline core/src/main/res/values-kab/strings.xml:2]: what is this?

File: core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkTestHelpers.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package org.kiwix.kiwixmobile.core.bookmark.viewmodel
+
+import org.kiwix.kiwixmobile.core.bookmark.adapter.BookmarkItem
+
+// dateFormat = d MMM yyyy
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkTestHelpers.kt:5]: comment not needed

File: core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkTestHelpers.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+//             5 Jul 2020
+fun bookmark(
+  bookmarkTitle: String = "bookmarkTitle",
+  isSelected: Boolean = false,
+  id: Long = 2
+): BookmarkItem {
+  return BookmarkItem(
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkTestHelpers.kt:12]: perhaps you don't need it for the tests but the style is to add a default parameter for every constructor param

File: core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkViewModelTest.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  private val sharedPreferenceUtil: SharedPreferenceUtil = mockk()
+
+  private lateinit var viewModel: BookmarkViewModel
+  private val testScheduler = TestScheduler()
+
+  private val itemsFromDb: PublishProcessor<List<BookmarkItem>> =
+    PublishProcessor.create()
+
+  init {
+    setScheduler(testScheduler)
```
Comment (macgills) [inline core/src/test/java/org/kiwix/kiwixmobile/core/bookmark/viewmodel/BookmarkViewModelTest.kt:65]: you don't need testScheduler so delete.

you do however need the IO scheduler to run on `trampoline` ie sequentially so you can just say

`setScheduler(Schedulers.trampoline())`

**Comment by macgills** [comment]
So while this does adhere to MVI the duplication across the history/bookmark activity/viewModels/State/Actions/SideEffects is extreme.

I am of 2 minds:
1) I can accept this and open a ticket for these classes to share their functionality
2) In this PR you can share the functionality

I'll leave this up to you how to proceed, this PR might be a welcome distraction & a palate cleanse from Navigation or it could be a nuisance distracting you.

File: core/src/main/res/values-kab/strings.xml (hunk 3 above)
Comment (Frans-Lukas) [reply core/src/main/res/values-kab/strings.xml:2]: Good question..

**Comment by Frans-Lukas** [comment]
It's definitely something nice to come back to after battling navigation. And it feels nice to be done with this, so I'll make sure they share the side effects!

File: core/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -28,7 +28,7 @@
       android:name=".search.SearchActivity"
       android:windowSoftInputMode="adjustResize" />
 
-    <activity android:name=".bookmark.BookmarksActivity" />
+    <activity android:name="org.kiwix.kiwixmobile.core.page.bookmark.BookmarksActivity" />
```
Comment (macgills) [inline core/src/main/AndroidManifest.xml:31]: We should still be able to keep a relative path, alt enter and see if you get an option or just play around and delete a few segments

File: core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDao.kt [hunk 8]
```diff
@@ -21,24 +21,15 @@ import io.objectbox.Box
...
 
 class NewBookmarksDao @Inject constructor(val box: Box<BookmarkEntity>) {
-  fun getBookmarks(
-    fromCurrentBook: Boolean,
-    zimFileReader: ZimFileReader?
-  ) = box.query {
-    if (fromCurrentBook) {
-      equal(BookmarkEntity_.zimName, zimFileReader?.name ?: "")
-    }
-    order(BookmarkEntity_.bookmarkTitle)
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/dao/NewBookmarksDao.kt:39]: they used to be ordered by title with this query, we may wish to restore that functionality

File: core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.java [hunk 9]
```diff
@@ -46,7 +46,7 @@
 
   Completable clearHistory();
 
-  Single<List<BookmarkItem>> getBookmarks(boolean showFromCurrentBook);
+  Flowable<List<BookmarkItem>> getBookmarks(boolean showFromCurrentBook);
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/data/DataSource.java:49]: I imagine this boolean is pointless

File: core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/BookmarksActivity.kt [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.Filter
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemClick
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.OnItemLongClick
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteButton
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedDeleteSelectedPages
+import org.kiwix.kiwixmobile.core.page.viewmodel.Action.UserClickedShowAllToggle
+import org.kiwix.kiwixmobile.core.utils.SimpleTextListener
+import javax.inject.Inject
+
+class BookmarksActivity : OnItemClickListener, BaseActivity() {
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/BookmarksActivity.kt:41]: This activity duplicates too much of what is in HistoryActivity.

Have you much experience with generics? Essentially I want to see an abstract base class

```
<T:Page>PageActivity<T>
```

and implementations
```
HistoryActivity: PageActivity<HistoryItem>{ }
BookmarkActivity: PageActivity<BookmarkItem>{ }
```
and in the implementations they will only contain the parts that absolutely have to differ.

A similar methodology may be applied to ViewModel and State.

File: core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/viewmodel/effects/DeleteBookmarkItems.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+data class DeleteBookmarkItems(
+  private val state: BookmarkState,
+  private val bookmarksDao: NewBookmarksDao
+) : SideEffect<Unit> {
+  override fun invokeWith(activity: AppCompatActivity) {
+    if (state.isInSelectionState) {
+      bookmarksDao.deleteBookmarks(state.bookmarks.filter(BookmarkItem::isSelected))
+    } else {
+      bookmarksDao.deleteBookmarks(state.bookmarks)
+      ShowToast(R.string.all_bookmarks_cleared)
```
Comment (macgills) [inline core/src/main/java/org/kiwix/kiwixmobile/core/page/bookmark/viewmodel/effects/DeleteBookmarkItems.kt:38]: ShowToast is a side effect, this should do nothing, you can just use the activityExtension `.toast()`

**Comment by macgills** [comment]
There is a conflict here so CI will not run

**Comment by Frans-Lukas** [comment]
I'm not super happy with this PageActivity solution. Still feels like there is lots that could be shared between the classes. Also, I couldn't figure out how to make it good-looking as a generic? There are many things individual to History and Bookmark like the adapter, different view names/ids, and viewmodel. Maybe all of them should be made generic as well? 

I am starting to think this should be a separate ticket (as it would be nice to be done with bookmarks.).

**Comment by macgills** [comment]
@Frans-Lukas Thats grand, up to yourself how you want to proceed. If we are doing no generification in this PR then I would want to see the attempts reveresed back to the state where Bookmark/History were completely distinct, not really a fan of merging something partway done

**Pull request merged**